### PR TITLE
Upgrade with new payment information

### PR DESCRIPTION
**Pull request by salbertson**
Prompt user for their payment info when their upgrade fails.

### CODE REVIEW

**Comment by salbertson** [comment]
[When a card is expired and user tries to upgrade the plan, we should surface the error from Stripe in the UI](https://trello.com/c/IlCA1XGF/831-when-a-card-is-expired-and-user-tries-to-upgrade-the-plan-we-should-surface-the-error-from-stripe-in-the-ui)

**Comment by salbertson** [comment]
[Users are unable to re-activate Hound if they had let a subscription in the past expire.](https://trello.com/c/hFmQeb47/833-users-are-unable-to-re-activate-hound-if-they-had-let-a-subscription-in-the-past-expire)

File: app/javascript/components/NotifyTierChange/components/UpgradeSubscriptionLink.jsx [hunk 1]
```diff
@@ -33,19 +39,42 @@ export default class UpgradeSubscriptionLink extends React.Component {
...
       allowRememberMe: false
     });
   }
 
+  handleFailedUpgrade(response) {
+    const { intercom } = this.props;
+
+    const subscriptionFailed = response &&
+      response.errors &&
+      response.errors.includes("There was an issue creating the subscription")
```
Comment (gylaz) [inline app/javascript/components/NotifyTierChange/components/UpgradeSubscriptionLink.jsx]: Semicolon at the end?

File: app/javascript/components/NotifyTierChange/components/UpgradeSubscriptionLink.jsx [hunk 2]
```diff
@@ -33,19 +39,42 @@ export default class UpgradeSubscriptionLink extends React.Component {
...
+
+    if (subscriptionFailed) {
+      this.showCreditCardForm();
+    } else {
+      this.setState({ disabled: false });
+
+      intercom(
+        "showNewMessage",
+        "I cannot upgrade and activate my repo. Please help!"
```
Comment (gylaz) [inline app/javascript/components/NotifyTierChange/components/UpgradeSubscriptionLink.jsx]: Does this start a thread conversation? Does the user know this is being sent on their behalf? Would it be surprising to the user?

File: spec/services/repo_subscriber_spec.rb [hunk 3]
```diff
@@ -22,14 +22,45 @@
...
+          RepoSubscriber.subscribe(repo, user, nil)
 
           expect(subscription_request).to have_been_requested
           expect(subscription_update_request).to have_been_requested
           expect(update_request).not_to have_been_requested
           expect(repo.subscription.stripe_subscription_id).
             to eq(stripe_subscription_id)
         end
+
+        it "creates a new Stripe subscription with new card, and repo subscription" do
```
[the remark this answers is not in the document]
Comment (gylaz) [reply spec/services/repo_subscriber_spec.rb]: Maybe remove a few words? Something like:
```
creates a Stripe and repo subscriptions with new card
```

File: spec/services/repo_subscriber_spec.rb [hunk 4]
```diff
@@ -22,14 +22,45 @@
...
+          repo = create(:repo, private: true)
+          user = create(
+            :user,
+            stripe_customer_id: stripe_customer_id,
```
Comment (gylaz) [inline spec/services/repo_subscriber_spec.rb]: Where does `stripe_customer_id` variable come from?

File: spec/services/repo_subscriber_spec.rb [hunk 5]
```diff
@@ -22,14 +22,45 @@
...
+            repos: [repo],
+          )
+          stub_customer_find_request
+          update_request = stub_customer_update_request
```
Comment (gylaz) [inline spec/services/repo_subscriber_spec.rb]: What's the difference between this one and `customer_update_request`?
Do we even need this assertion? If `customer_update_request` i called once, then we know that no other `customer update` requests were made, right?

File: app/javascript/components/NotifyTierChange/components/UpgradeSubscriptionLink.jsx (hunk 2 above)
Comment (salbertson) [reply app/javascript/components/NotifyTierChange/components/UpgradeSubscriptionLink.jsx]: It will open the Intercom messenger and pre-fill the message. The user will have a chance to edit the message, or include info, and click "send".

File: spec/services/repo_subscriber_spec.rb (hunk 4 above)
Comment (salbertson) [reply spec/services/repo_subscriber_spec.rb]: Looks like it's a spec helper method.

https://github.com/houndci/hound/blob/986aec32a1b3aef5df38aea079b5bcd05168808d/spec/support/helpers/stripe_api_helper.rb#L2-L4

File: spec/services/repo_subscriber_spec.rb (hunk 5 above)
Comment (salbertson) [reply spec/services/repo_subscriber_spec.rb]: Good call, we don't need this, you're right.

File: app/controllers/subscriptions_controller.rb [hunk 6]
```diff
@@ -72,4 +67,34 @@ def update_email
...
+      else
+        activator.deactivate
+        render_error("There was an issue creating the subscription")
+      end
+    else
+      render_error("There was an issue activating the repo")
+    end
+  end
+
+  def render_error(error)
```
[the remark this answers is not in the document]
Comment (gylaz) [reply app/controllers/subscriptions_controller.rb]: Should we change this to 15?

File: app/jobs/accept_github_invitation_job.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+class AcceptGitHubInvitationJob < ApplicationJob
+  queue_as :high
+
+  def perform(repo_name)
+    return if hound_github.repository?(repo_name)
```
Comment (gylaz) [inline app/jobs/accept_github_invitation_job.rb]: We've been avoiding these early returns, instead favoring:
```rb
unless hound_github.repository?(repo_name)
  hound_github.accept_invitation(repo_name)
end
```

Thoughts?

File: app/jobs/accept_github_invitation_job.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+class AcceptGitHubInvitationJob < ApplicationJob
```
Comment (gylaz) [inline app/jobs/accept_github_invitation_job.rb]: This doesn't seem related to this PR.

File: spec/jobs/accept_github_invitation_job_spec.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        expect(github).not_to have_received(:accept_invitation)
+      end
+    end
+
+    context "when Hound doesn't have access to the repo" do
+      it "accepts the invitation" do
+        github = double(GitHubApi, repository?: false, accept_invitation: nil)
+        allow(GitHubApi).to receive(:new).and_return(github)
+
+        job = subject.perform("hound/test")
```
[the remark this answers is not in the document]
Comment (salbertson) [reply spec/jobs/accept_github_invitation_job_spec.rb]: 🦅 👁

File: app/jobs/accept_github_invitation_job.rb (hunk 7 above)
Comment (salbertson) [reply app/jobs/accept_github_invitation_job.rb]: I don't mind the "guard" style, but I'm cool with it if that's been the convention. Any good reason other than convention?

File: app/jobs/accept_github_invitation_job.rb (hunk 8 above)
Comment (salbertson) [reply app/jobs/accept_github_invitation_job.rb]: You got me. Should I pull this into a separate PR?

File: app/jobs/accept_github_invitation_job.rb (hunk 7 above)
Comment (gylaz) [reply app/jobs/accept_github_invitation_job.rb]: > Any good reason other than convention?

The only good argument that I've heard is one-liners hide complexity:

It's more explicit that there is clearly branching logic/complexity when using `if/else`. This really comes into play if you have multiple early `return` lines -- looks simple, but really there's a bit of complexity to the method.

File: app/jobs/accept_github_invitation_job.rb (hunk 8 above)
Comment (salbertson) [reply app/jobs/accept_github_invitation_job.rb]: https://github.com/houndci/hound/pull/1501

File: spec/controllers/subscriptions_controller_spec.rb [hunk 10]
```diff
@@ -101,6 +102,7 @@
           email: "TEST_USER_EMAIL",
           repos: repos,
           token: "TEST_USER_TOKEN",
+          username: "jimtom",
```
Comment (gylaz) [inline spec/controllers/subscriptions_controller_spec.rb]: Jimtom lives!
